File: client/src/main/scala/stasis/client/encryption/secrets/UserLocalEncryptionSecret.scala
package stasis.client.encryption.secrets

import scala.concurrent.Future

import org.apache.pekko.stream.Materializer
import org.apache.pekko.util.ByteString

import stasis.client.encryption.Aes
import stasis.shared.model.devices.Device
import stasis.shared.model.users.User
import stasis.shared.secrets.SecretsConfig

final case class UserLocalEncryptionSecret(
  user: User.Id,
  iv: ByteString,
  private val key: ByteString
)(implicit target: SecretsConfig)
    extends Secret {

  def encryptDeviceSecret(
    secret: DeviceSecret
  )(implicit mat: Materializer): Future[ByteString] =
    secret.encrypted(Aes.encryption(key, iv))

  def decryptDeviceSecret(
    device: Device.Id,
    encryptedSecret: ByteString
  )(implicit mat: Materializer): Future[DeviceSecret] =
    DeviceSecret.decrypted(
      user = user,
      device = device,
      encryptedSecret = encryptedSecret,
      Aes.decryption(key, iv)
    )
}

object UserLocalEncryptionSecret {
  def apply(
    user: User.Id,
    iv: ByteString,
    key: ByteString
  )(implicit target: SecretsConfig): UserLocalEncryptionSecret =
    new UserLocalEncryptionSecret(user, iv, key)
}

File: client-ios/lib/Sources/StasisClientLib/Encryption/Secrets/UserLocalEncryptionSecret.swift
import Foundation

public struct UserLocalEncryptionSecret: Secret, Equatable {
    public let user: UserId
    public let iv: Data
    private let key: Data
    public let target: SecretConfig

    public init(user: UserId, iv: Data, key: Data, target: SecretConfig) {
        self.user = user
        self.iv = iv
        self.key = key
        self.target = target
    }

    public func encryptDeviceSecret(_ secret: DeviceSecret) throws -> Data {
        try secret.encrypted { plaintext in
            try Aes.encrypt(plaintext: plaintext, key: key, iv: iv)
        }
    }

    public func decryptDeviceSecret(device: DeviceId, encryptedSecret: Data) throws -> DeviceSecret {
        try DeviceSecret.decrypted(
            user: user,
            device: device,
            encryptedSecret: encryptedSecret,
            decryptionStage: { ciphertext in
                try Aes.decrypt(ciphertext: ciphertext, key: key, iv: iv)
            },
            target: target
        )
    }
}
